File: zoobin.py
import sys
from collections import deque

sys.setrecursionlimit(2000)

def compose_perm(p1, p2, size):
    res = [0] * (size + 1)
    for i in range(1, size + 1):
        res[i] = p1[p2[i]]
    return res

def to_tuple_key(perm):
    return tuple(perm[1:])

def main():
    try:
        try:
            E_line = input()
            if not E_line:
                return
            edge_count = int(E_line)
        except EOFError:
            return

        edges_A = []
        edges_B = []
        max_vertex = 0

        for _ in range(edge_count):
            x, y = map(int, input().split())
            if x > y:
                x, y = y, x
            edges_A.append((x, y))
            max_vertex = max(max_vertex, y)

        for _ in range(edge_count):
            x, y = map(int, input().split())
            if x > y:
                x, y = y, x
            edges_B.append((x, y))
            max_vertex = max(max_vertex, y)

        n = max_vertex
        graphA = [[0] * (n + 1) for _ in range(n + 1)]
        graphB = [[0] * (n + 1) for _ in range(n + 1)]

        for u, v in edges_A:
            graphA[u][v] = graphA[v][u] = 1
        for u, v in edges_B:
            graphB[u][v] = graphB[v][u] = 1

        degA = [0] * (n + 1)
        degB = [0] * (n + 1)
        for i in range(1, n + 1):
            for j in range(1, n + 1):
                degA[i] += graphA[i][j]
                degB[i] += graphB[i][j]

        node_order = list(range(1, n + 1))
        node_order.sort(key=lambda x: (-degA[x], x))
        used_nodes = [0] * (n + 1)
        perm_map = [0] * (n + 1)

        def dfs_iso(idx):
            if idx == len(node_order):
                return True
            v = node_order[idx]
            for w in range(1, n + 1):
                if used_nodes[w]:
                    continue
                if degA[v] != degB[w]:
                    continue
                valid = True
                for k in range(idx):
                    u = node_order[k]
                    if graphA[v][u] and perm_map[u] != 0 and not graphB[w][perm_map[u]]:
                        valid = False
                        break
                    if not graphA[v][u] and perm_map[u] != 0 and graphB[w][perm_map[u]]:
                        valid = False
                        break
                if not valid:
                    continue
                used_nodes[w] = 1
                perm_map[v] = w
                if dfs_iso(idx + 1):
                    return True
                used_nodes[w] = 0
                perm_map[v] = 0
            return False

        dfs_iso(0)

        inverse_perm = [0] * (n + 1)
        for i in range(1, n + 1):
            if perm_map[i] != 0:
                inverse_perm[perm_map[i]] = i

        adj_list = [[] for _ in range(n + 1)]
        for i in range(1, n + 1):
            for j in range(1, n + 1):
                if graphA[i][j]:
                    adj_list[i].append(j)
                    adj_list[j].append(i)

        all_cycles = []
        visited = [0] * (n + 1)
        path = []

        def dfs_cycle(start, u):
            for v in adj_list[u]:
                if v < start:
                    continue
                if v == start and len(path) >= 3:
                    all_cycles.append(list(path))
                    continue
                if not visited[v] and v > start:
                    visited[v] = 1
                    path.append(v)
                    dfs_cycle(start, v)
                    path.pop()
                    visited[v] = 0

        for s in range(1, n + 1):
            visited = [0] * (n + 1)
            path = [s]
            visited[s] = 1
            dfs_cycle(s, s)

        generators = []
        identity = list(range(n + 1))

        for cycle in all_cycles:
            k = len(cycle)
            if k == 0:
                continue

            forward = list(identity)
            for i in range(k):
                a = cycle[i]
                b = cycle[(i + 1) % k]
                forward[a] = b
            generators.append(forward)

            backward = list(identity)
            for i in range(k):
                a = cycle[i]
                b = cycle[(i - 1 + k) % k]
                backward[a] = b
            generators.append(backward)

        identity_perm = list(range(n + 1))
        if inverse_perm == identity_perm:
            print(0, end="")
            return

        q = deque()
        dist_map = {}
        key_identity = to_tuple_key(identity_perm)
        q.append(identity_perm)
        dist_map[key_identity] = 0
        result = -1

        while q:
            curr = q.popleft()
            key_curr = to_tuple_key(curr)
            steps = dist_map[key_curr]

            for gen in generators:
                nxt = compose_perm(curr, gen, n)
                nxt_key = to_tuple_key(nxt)

                if nxt_key not in dist_map:
                    dist_map[nxt_key] = steps + 1
                    if nxt == inverse_perm:
                        result = steps + 1
                        print(result, end="")
                        return
                    q.append(nxt)
        print(result, end="")

    except EOFError:
        pass

main()

File: zoobin.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> compose_perm(const vector<int>& p1, const vector<int>& p2, int size) {
    vector<int> res(size + 1);
    for (int i = 1; i <= size; ++i)
        res[i] = p1[p2[i]];
    return res;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int edge_count;
    if (!(cin >> edge_count)) return 0;

    vector<pair<int, int>> edges_A, edges_B;
    int max_vertex = 0;

    for (int i = 0; i < edge_count; ++i) {
        int x, y;
        cin >> x >> y;
        if (x > y) swap(x, y);
        edges_A.emplace_back(x, y);
        max_vertex = max(max_vertex, y);
    }

    for (int i = 0; i < edge_count; ++i) {
        int x, y;
        cin >> x >> y;
        if (x > y) swap(x, y);
        edges_B.emplace_back(x, y);
        max_vertex = max(max_vertex, y);
    }

    int n = max_vertex;
    vector<vector<int>> graphA(n + 1, vector<int>(n + 1));
    vector<vector<int>> graphB(n + 1, vector<int>(n + 1));

    for (auto [u, v] : edges_A) graphA[u][v] = graphA[v][u] = 1;
    for (auto [u, v] : edges_B) graphB[u][v] = graphB[v][u] = 1;

    vector<int> degA(n + 1), degB(n + 1);
    for (int i = 1; i <= n; ++i)
        for (int j = 1; j <= n; ++j) {
            degA[i] += graphA[i][j];
            degB[i] += graphB[i][j];
        }

    vector<int> node_order(n);
    iota(node_order.begin(), node_order.end(), 1);
    sort(node_order.begin(), node_order.end(), [&](int x, int y) {
        return make_pair(-degA[x], x) < make_pair(-degA[y], y);
    });

    vector<int> used_nodes(n + 1), perm_map(n + 1);
    function<bool(int)> dfs_iso = [&](int idx) {
        if (idx == node_order.size()) return true;
        int v = node_order[idx];
        for (int w = 1; w <= n; ++w) {
            if (used_nodes[w] || degA[v] != degB[w]) continue;
            bool valid = true;
            for (int k = 0; k < idx; ++k) {
                int u = node_order[k];
                if ((graphA[v][u] && perm_map[u] && !graphB[w][perm_map[u]]) ||
                    (!graphA[v][u] && perm_map[u] && graphB[w][perm_map[u]])) {
                    valid = false;
                    break;
                }
            }
            if (!valid) continue;
            used_nodes[w] = 1;
            perm_map[v] = w;
            if (dfs_iso(idx + 1)) return true;
            used_nodes[w] = 0;
            perm_map[v] = 0;
        }
        return false;
    };

    dfs_iso(0);

    vector<int> inverse_perm(n + 1);
    for (int i = 1; i <= n; ++i)
        if (perm_map[i]) inverse_perm[perm_map[i]] = i;

    vector<vector<int>> adj_list(n + 1);
    for (int i = 1; i <= n; ++i)
        for (int j = 1; j <= n; ++j)
            if (graphA[i][j]) adj_list[i].push_back(j);

    vector<vector<int>> all_cycles;
    vector<int> visited(n + 1), path;

    function<void(int, int)> dfs_cycle = [&](int start, int u) {
        for (int v : adj_list[u]) {
            if (v < start) continue;
            if (v == start && path.size() >= 3) {
                all_cycles.push_back(path);
                continue;
            }
            if (!visited[v] && v > start) {
                visited[v] = 1;
                path.push_back(v);
                dfs_cycle(start, v);
                path.pop_back();
                visited[v] = 0;
            }
        }
    };

    for (int s = 1; s <= n; ++s) {
        fill(visited.begin(), visited.end(), 0);
        path = {s};
        visited[s] = 1;
        dfs_cycle(s, s);
    }

    vector<vector<int>> generators;
    vector<int> identity(n + 1);
    iota(identity.begin(), identity.end(), 0);

    for (const auto& cycle : all_cycles) {
        int k = cycle.size();
        if (k == 0) continue;

        vector<int> forward = identity;
        for (int i = 0; i < k; ++i)
            forward[cycle[i]] = cycle[(i + 1) % k];
        generators.push_back(forward);

        vector<int> backward = identity;
        for (int i = 0; i < k; ++i)
            backward[cycle[i]] = cycle[(i - 1 + k) % k];
        generators.push_back(backward);
    }

    if (inverse_perm == identity) {
        cout << 0;
        return 0;
    }

    queue<vector<int>> q;
    map<vector<int>, int> dist_map;
    q.push(identity);
    dist_map[identity] = 0;

    while (!q.empty()) {
        auto curr = q.front(); q.pop();
        int steps = dist_map[curr];

        for (const auto& gen : generators) {
            auto nxt = compose_perm(curr, gen, n);
            if (!dist_map.count(nxt)) {
                dist_map[nxt] = steps + 1;
                if (nxt == inverse_perm) {
                    cout << steps + 1;
                    return 0;
                }
                q.push(nxt);
            }
        }
    }

    cout << -1;
    return 0;
}